File: solutions/typescript/resistor-color-trio/2/resistor-color-trio.ts
export function decodedResistorValue(arr: string[]): string {
  const vals: string = `${colorCode(arr[0])}${colorCode(arr[1])}`;
  const nzero: number = colorCode(arr[2]);
  if (nzero > 0) {
    let zeros = '0';
    zeros = zeros.repeat(nzero);
    return addUnits(Number(vals+zeros));
  }
  return addUnits(Number(vals));
}

const colorCode = (color: string): number => {
  return COLORS.indexOf(color);
}

const addUnits = (val: number): string => {
  const giga = 1000000000;
  const mega = 1000000;
  const kilo = 1000;
  if (val > giga) {
    return `${val / giga} gigaohms`;
  }
if (val > mega) {
    return `${val / mega} megaohms`;
  }
  if (val > kilo) {
    return `${val / kilo} kiloohms`;
  }
  return `${val} ohms`;
}
const COLORS =
  [
    'black',
    'brown',
    'red',
    'orange',
    'yellow',
    'green',
    'blue',
    'violet',
    'grey',
    'white',
  ]

File: solutions/zig/resistor-color-trio/1/resistor_color_trio.zig
const std = @import("std");
const mem = std.mem;

pub const ColorBand = enum(u8) { black, brown, red, orange, yellow, green, blue, violet, grey, white };
const kilo = 1_000;
const mega = 1_000_000;
const giga = 1_000_000_000;

pub fn label(allocator: mem.Allocator, colors: []const ColorBand) mem.Allocator.Error![]u8 {
    const num: u64 = ((@intFromEnum(colors[0]) * 10) + @intFromEnum(colors[1])) * std.math.pow(u64, 10, @intFromEnum(colors[2]));
    if (num >= giga) {
        return std.fmt.allocPrint(allocator, "{d} gigaohms", .{@as(f64, @floatFromInt(num)) / giga});
    }
    if (num >= mega) {
        return std.fmt.allocPrint(allocator, "{d} megaohms", .{@as(f64, @floatFromInt(num)) / mega});
    }
    if (num >= kilo) {
        return std.fmt.allocPrint(allocator, "{d} kiloohms", .{@as(f64, @floatFromInt(num)) / kilo});
    }
    return std.fmt.allocPrint(allocator, "{d} ohms", .{num});
}
